File: example3/main.py
from wasmtime import Store, Module, Instance, Func, FuncType, ValType # type: ignore

def print_result(x):
    print(f"Result found: {x}")

store = Store()
module = Module.from_file(store.engine, 'sum.wasm')
print_result_type = FuncType([ValType.i32()], [])
print_result_wrapper = Func(store, print_result_type, print_result)
imports = [print_result_wrapper]
instance = Instance(store, module, imports)
exports = instance.exports(store)
sum = exports["sum"]
result = sum(store, 20, 22)
print(result)

File: example3/main.rs
use wasmtime::*;

fn print_result(x: i32) {
    println!("Host function: {}", x);
}

fn main() -> Result<()> {
    let engine = Engine::default();
    let module = Module::from_file(&engine, "sum.wasm")?;
    let mut store = Store::new(&engine, ());

    let mut linker: Linker<()> = Linker::new(&engine);
    linker.func_wrap("env", "print_result", print_result)?;

    let instance = linker.instantiate(&mut store, &module)?;
    
    let sum = instance.get_typed_func::<(i32, i32), i32>(&mut store, "sum")?;
    let result= sum.call(&mut store, (20, 22))?;
    println!("Guest function: {result}");

    Ok(())
}
